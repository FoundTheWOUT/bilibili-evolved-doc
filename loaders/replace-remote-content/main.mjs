/**
 * 替换 <remote /> 为 src 里内容
 * Author: FoundTheWOUT
 * */

import { remark } from "remark";
import { visit } from "unist-util-visit";
import { toHtml } from "hast-util-to-html";
import { fromHtml } from "hast-util-from-html";
import { fromMarkdown } from "mdast-util-from-markdown";
import { remove } from "unist-util-remove";
import fetch from "node-fetch";

// TODO, check valid jsx.
const isJsxNode = (node) => {
  // if is tsx component, should not handle it.
  if (
    node.value[0] == "<" &&
    (/[A-Z]/.test(node.value[1]) || /[A-Z]/.test(node.value[2]))
  ) {
    return true;
  }
  return false;
};

const normalizeHTML = (tree) => {
  tree = remove(tree, { tagName: "style" });
  if (!tree) return;
  tree = remove(tree, { type: "comment" });
  if (!tree) return;
  return toHtml(tree, {
    closeSelfClosing: true,
  });
};

const normalizeContentPlugin = () => {
  return (tree) => {
    visit(tree, (node) => {
      if (!node.value) return;
      // ! only modify img
      if (
        node.type == "html" &&
        (node.value.includes("<img") ||
          node.value.includes("<!-") ||
          node.value.includes("<style"))
      ) {
        if (isJsxNode(node)) return;
        const tree = fromHtml(node.value, { fragment: true }) ?? "";
        node.value = normalizeHTML(tree);
      }
      if (node.type == "code" && node.lang == "powershell") {
        node.lang = "shell";
      }
    });
  };
};

const handleSectionSelect = (tree, options) => {
  if (options.section.length == 0) return tree;
  let depth = -1; // depth != -1 meaning should retained the node.
  let index = 0;
  while (index < tree.children.length) {
    // select the current node
    const node = tree.children[index];
    switch (node.type) {
      case "heading":
        /**
         * for example, a md:
         * ### Foo <- selected section
         * Bar
         * ## Jack <- drop content from here
         * 19
         */
        if (depth !== -1 && node.depth <= depth) {
          depth = -1;
        }
        for (let child of node.children) {
          if (options.section.includes(child.value)) {
            depth = node.depth; // record the depth
            index++;
          } else if (depth !== -1) {
            index++;
          } else {
            tree.children.splice(index, 1);
          }
        }
        break;
      default:
        if (depth == -1) {
          // if not collecting item, remove it from children directly.
          tree.children.splice(index, 1);
        } else {
          index++;
        }
        break;
    }
  }
  return tree;
};
// TODO: write to local.
const cache = new Map();
const fetchContentPlugin = (options = {}) => {
  const { section, dry } = options;
  return async (tree) => {
    if (dry && section) {
      handleSectionSelect(tree, options);
      return;
    }
    const promises = [];
    let preFragmentLength = 0;
    // 1. select mdxJsxFlowElement type, and get content from url.
    // mdast -> hast for later use
    visit(tree, "html", (node, offset, parent) => {
      if (isJsxNode(node)) return;
      const hast = fromHtml(node.value, { fragment: true });
      if (
        hast.children &&
        hast.children[0] &&
        hast.children[0].tagName == "remote"
      ) {
        const hastNode = hast.children[0];
        let url = null;
        let sections = [];
        for (let [attrName, value] of Object.entries(hastNode.properties)) {
          if (attrName == "src") {
            url = value;
          }
          if (/section/g.test(attrName)) {
            sections.push(value);
          }
        }
        if (url) {
          promises.push(() => {
            console.log("fetching:", url);
            const encodedUrl = encodeURI(url);
            const cacheContent = cache.get(encodedUrl);
            let _p = null;
            if (cacheContent) {
              console.log("url:", url, "hit cache!");
              _p = Promise.resolve(cacheContent);
            } else {
              _p = fetch(url)
                .then((res) => res.text())
                .then((text) => {
                  cache.set(encodedUrl, text);
                  return text;
                });
            }

            return _p
              .then((text) => {
                console.log("compiling markdown to mdast...");
                return fromMarkdown(text);
              })
              .then((mdast) => {
                console.log("selecting section...");
                return handleSectionSelect(mdast, { section: sections });
              })
              .then((mdast) => {
                parent.children.splice(
                  offset + preFragmentLength,
                  1,
                  ...mdast.children
                );
                preFragmentLength += mdast.children.length - 1;
              })
              .catch((error) => {
                console.log("fetch content error:", error);
              });
          });
        }
      }
    });

    for (let p of promises) {
      try {
        await p();
      } catch (error) {
        switch (process.env.NODE_ENV) {
          case "development":
            console.error(error);
          default: {
            throw Error(error);
            process.exit(0);
          }
        }
      }
    }
  };
};

/**
 *
 * @param {string} content - .mdx or .md file
 * @param {(_:undefined,value:string)=>void} callback
 * @param {{dry:boolean}|undefined}options - when passing the dry options, the loader would not find <remote>, but treat it like a file which after fetching content from url.
 * @returns
 */
export async function loader(content, callback, options) {
  remark()
    .use(fetchContentPlugin, options)
    .use(normalizeContentPlugin)
    .process(content)
    .then((res) => {
      callback(null, res.toString());
    });
}

export const loaderPromise = (template, options) =>
  new Promise((resolve, reject) => {
    loader(
      template,
      (_, value) => {
        resolve(value);
      },
      options
    );
  });
