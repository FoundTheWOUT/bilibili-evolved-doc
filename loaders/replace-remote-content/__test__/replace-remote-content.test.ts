import { loaderPromise as __replaceFun } from "../main.mjs";
import { describe, it, expect, beforeAll } from "vitest";
import { readFile } from "fs/promises";
import handler from "serve-handler";
import { createServer } from "http";

// first, host
const TEST_PORT = 5500;
beforeAll(async () => {
  const server = createServer((request, response) => {
    return handler(request, response, {
      // ! fix the path
      rewrites: [
        {
          source: "__test__/:name",
          destination: "loaders/replace-remote-content/__test__/:name",
        },
      ],
    });
  });
  server.listen(TEST_PORT, () => {
    console.log(`Running at http://localhost:${TEST_PORT}`);
  });

  return async () => {
    server.close();
  };
});

describe("Unit Test replace RemoteContent", () => {
  it("can fetch remote content", async () => {
    const content = '<remote src="http://127.0.0.1:5500/__test__/example.md"/>';

    const res = await __replaceFun(content);
    expect(res).toMatchSnapshot();
  });

  it("fetch remote content, and select two section", async () => {
    const content =
      '<remote src="http://127.0.0.1:5500/__test__/example.md" section="Sugaring time begins in February and lasts for two to six weeks" section-two="Twice a day, and some days more often"/>';

    const res = await __replaceFun(content);
    expect(res).toMatchSnapshot();
  });

  it("cache hit", async () => {
    const content = `<remote src="http://127.0.0.1:5500/__test__/example.md" section="Sugaring time begins in February and lasts for two to six weeks"/>

## Foo
Bar

<remote src="http://127.0.0.1:5500/__test__/example.md" section="Sugaring time begins in February and lasts for two to six weeks"/>

<remote src="http://127.0.0.1:5500/__test__/example.md" section="Sugaring time begins in February and lasts for two to six weeks"/>
`;

    const res = await __replaceFun(content);
    expect(res).toMatchSnapshot();
  });

  it("html tag", async () => {
    const template = `# 合集包

合集包提供了批量的功能安装链接, 方便一次性安装大量功能.

#### 简洁至上

简化各种多余界面元素, 专注于内容本身.

包含以下功能:
删除广告, 删除直播水印, 删除视频弹窗, 禁用特殊弹幕样式, 简化评论区, 简化直播间, 简化首页, 自动收起直播侧栏, 隐藏视频推荐, 隐藏直播推荐, 隐藏视频标题层, 自动隐藏侧栏

<details>
<summary><strong>jsDelivr Stable</strong></summary>
https://cdn.jsdelivr.net/gh/the1812/Bilibili-Evolved@v2/registry/dist/components/utils/remove-promotions.js
https://cdn.jsdelivr.net/gh/the1812/Bilibili-Evolved@v2/registry/dist/components/live/remove-watermark.js
https://cdn.jsdelivr.net/gh/the1812/Bilibili-Evolved@v2/registry/dist/components/video/player/remove-popup.js
https://cdn.jsdelivr.net/gh/the1812/Bilibili-Evolved@v2/registry/dist/components/style/special-danmaku.js
https://cdn.jsdelivr.net/gh/the1812/Bilibili-Evolved@v2/registry/dist/components/style/simplify/comments.js
https://cdn.jsdelivr.net/gh/the1812/Bilibili-Evolved@v2/registry/dist/components/style/simplify/live.js
https://cdn.jsdelivr.net/gh/the1812/Bilibili-Evolved@v2/registry/dist/components/style/simplify/home.js
https://cdn.jsdelivr.net/gh/the1812/Bilibili-Evolved@v2/registry/dist/components/live/side-bar.js
https://cdn.jsdelivr.net/gh/the1812/Bilibili-Evolved@v2/registry/dist/components/style/hide/video/related-videos.js
https://cdn.jsdelivr.net/gh/the1812/Bilibili-Evolved@v2/registry/dist/components/style/hide/video/recommended-live.js
https://cdn.jsdelivr.net/gh/the1812/Bilibili-Evolved@v2/registry/dist/components/style/hide/video/top-mask.js
https://cdn.jsdelivr.net/gh/the1812/Bilibili-Evolved@v2/registry/dist/components/style/auto-hide-sidebar.js
</details>

<remote src="http://127.0.0.1:5500/__test__/example.md" section="Sugaring time begins in February and lasts for two to six weeks"/>
`;
    const res = await __replaceFun(template);
    expect(res).toMatchSnapshot();
  });

  it("should return mdx properly", async () => {
    const mdx = (
      await readFile("loaders/replace-remote-content/__test__/example.mdx")
    ).toString();
    const res = await __replaceFun(mdx);
    expect(res).toMatchSnapshot();
  });

  it("html image self close", async () => {
    const template = `<remote src="http://127.0.0.1:5500/__test__/image-self-close.md"/>`;
    const res = await __replaceFun(template);
    expect(res).toMatchSnapshot();
  });

  it("should remove html comment", async () => {
    const template = `<remote src="http://127.0.0.1:5500/__test__/html-comment.md"/>`;
    const res = await __replaceFun(template);
    expect(res).eq("\n\n");
  });

  it("don't break mdast structure", async () => {
    const content =
      '<remote src="http://127.0.0.1:5500/__test__/changelog.md"/>';

    const res = (await __replaceFun(content)) as string;
    expect(res.replaceAll("\r", "")).toMatchSnapshot();
  });
});
