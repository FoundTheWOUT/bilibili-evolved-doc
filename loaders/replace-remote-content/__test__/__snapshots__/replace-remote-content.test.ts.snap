// Vitest Snapshot v1

exports[`Unit Test replace RemoteContent > cache hit 1`] = `
"### Sugaring time begins in February and lasts for two to six weeks

well into March. Weather permitting, Grandad always started tapping on Washington's Birthday. The sugaring season, as we called it, required thawing days and freezing nights to allow the sap to flow. Rain spoiled the sap and wind dried up the tap holes so it was a fair-weather industry. Grandad hand-drilled his tap holes using a No. 8 wood bit, drilling about 1/2 inch deep into the vein of a sugar maple tree. Into this hole, a wood spile would then be tapped and after drilling another, a pail was hung to collect the sap as it steadily dripped from the spile. The wooden spiles were about eight inches in length and three-quarters of an inch in diameter. Ours were made from elderberry as that scrub has a soft center that could be cleaned out to make a tube. Our maple grove was along the brook and the edge of the field about a have mile up from the house. There were approximately 200 sugar maples large enough to tap on Grandad's farm and the scars on the trunks of many of the older trees indicated the number of years they had been supplying sap.

## Foo

Bar

### Sugaring time begins in February and lasts for two to six weeks

well into March. Weather permitting, Grandad always started tapping on Washington's Birthday. The sugaring season, as we called it, required thawing days and freezing nights to allow the sap to flow. Rain spoiled the sap and wind dried up the tap holes so it was a fair-weather industry. Grandad hand-drilled his tap holes using a No. 8 wood bit, drilling about 1/2 inch deep into the vein of a sugar maple tree. Into this hole, a wood spile would then be tapped and after drilling another, a pail was hung to collect the sap as it steadily dripped from the spile. The wooden spiles were about eight inches in length and three-quarters of an inch in diameter. Ours were made from elderberry as that scrub has a soft center that could be cleaned out to make a tube. Our maple grove was along the brook and the edge of the field about a have mile up from the house. There were approximately 200 sugar maples large enough to tap on Grandad's farm and the scars on the trunks of many of the older trees indicated the number of years they had been supplying sap.

### Sugaring time begins in February and lasts for two to six weeks

well into March. Weather permitting, Grandad always started tapping on Washington's Birthday. The sugaring season, as we called it, required thawing days and freezing nights to allow the sap to flow. Rain spoiled the sap and wind dried up the tap holes so it was a fair-weather industry. Grandad hand-drilled his tap holes using a No. 8 wood bit, drilling about 1/2 inch deep into the vein of a sugar maple tree. Into this hole, a wood spile would then be tapped and after drilling another, a pail was hung to collect the sap as it steadily dripped from the spile. The wooden spiles were about eight inches in length and three-quarters of an inch in diameter. Ours were made from elderberry as that scrub has a soft center that could be cleaned out to make a tube. Our maple grove was along the brook and the edge of the field about a have mile up from the house. There were approximately 200 sugar maples large enough to tap on Grandad's farm and the scars on the trunks of many of the older trees indicated the number of years they had been supplying sap.
"
`;

exports[`Unit Test replace RemoteContent > can fetch remote content 1`] = `
"# Sugaring As I Remember It

## The end of winter once again brings with it the memories of maple sugaring

Back when the Highlands area was mostly farms and woods rather than housing developments as it is today, the manufacturing of maple sugar products was an excellent source of extra income for the local farmers.

## The tapping of the sugar maple was an important lesson learned by our forefathers from the Native Americans

as it provided many Colonial families their only source of sugar. Up until a few years ago, generations of the Post family had been making maple sugar on farms originally acquired by Peter and Abram Post in 1815 in the shadow of Bearfort Mountain. This was an art passed down from father to son unto the present generation. With the break up of the farms into home lots, all that remains are memories and, luckily, some of the equipment used through the years.

### Sugaring time begins in February and lasts for two to six weeks

well into March. Weather permitting, Grandad always started tapping on Washington's Birthday. The sugaring season, as we called it, required thawing days and freezing nights to allow the sap to flow. Rain spoiled the sap and wind dried up the tap holes so it was a fair-weather industry. Grandad hand-drilled his tap holes using a No. 8 wood bit, drilling about 1/2 inch deep into the vein of a sugar maple tree. Into this hole, a wood spile would then be tapped and after drilling another, a pail was hung to collect the sap as it steadily dripped from the spile. The wooden spiles were about eight inches in length and three-quarters of an inch in diameter. Ours were made from elderberry as that scrub has a soft center that could be cleaned out to make a tube. Our maple grove was along the brook and the edge of the field about a have mile up from the house. There were approximately 200 sugar maples large enough to tap on Grandad's farm and the scars on the trunks of many of the older trees indicated the number of years they had been supplying sap.

## The saphouse was a small

low building built into the ledge along Clinton Brook. Its walls were of stone and, having caught fire at least once, the building, I remember, had a tin roof so only the door was of wood. At the rear against the stone ledge was the fireplace with a dirt hearth across the whole width of the building. A wooden pole above the hearth rested at each end on the low side walls, blackened by age and smoke. From this, on wrought iron pot hooks, hung three large iron kettles and one brass kettle. The roof above the kettles was hinged like two large doors to let the smoke out and keep the rain from coming in. On calm days these doors could be laid completely open, back on the roof. On windy days they could b regulated so that the smoke did not blow back into the saphouse and into the boiling sap. To prevent sparks from going into the hemlock woods, a screen of tin was placed around the rear on the top of the ledge. Clinton Brook flowing by the door not only provided a picturesque setting but was essential for scouring the kettles, strainers and utensils after each boiling.

## Twice a day, and some days more often

the sap was collected from each tree and carried up to the saphouse. The kettles were filled and, with a good hot fire beneath, thirty pails of sap would boil down to about eight quarts of syrup in four to five hours. As the sap boiled down and thickened it was removed from the fire and strained and then placed in the brass kettle over a low fire for the last bit of boiling down to syrup. The syrup, as it came from the saphouse, was usually not quite thick enough so it was Grandma's job to boil it a little longer on the back of the old wood-burning kitchen stove. For a thicker cream sugar (we called it \\"slush\\") and for sugar cakes, additional boiling and beating were required with all hands taking a turn. The beating cooled the heavy syrup and as it cooled it became granular and almost white. Our reward was the pan to \\"lick\\" out and hot syrup poured out on the snow for taffy.

### My brothers, sisters, and I, as youngsters

were at the saphouse every chance we got during the sugaring season, playing in the brook and jumping from rock to rock. If you slipped and got wet, the fire was always handy. When we grew older we helped to gather the sap, split wood, and tend the fire. After Grandad's death, I did the whole bit myself for a few years.

## I especially remember the late thirties and early forties when the Newark Sunday News would do a picture story on Grandad and the sugaring.

This brought out the sightseers and every weekend there would be groups of people around asking questions and taking pictures. Some came back each year and many came to buy the maple sugar products. The more adventurous ones would hike with us to the falls to view the old Clinton Iron Furnace. Many a time we would pack a lunch and stay all day at the saphouse, when school was not in session, and the weather was nice. But usually, it was up the hill to the \\"big house\\" with Grandad for dinner and the noonday farm chores; then back to the saphouse till supper time. If it was a season with a good run of sap, Grandad would boil some down at night. That was the best time of all, sitting in the saphouse with the fire for warmth and the old man's stories of long ago to pass the hours.

<div>





\`\`\`shell
echo hi
\`\`\`
"
`;

exports[`Unit Test replace RemoteContent > don't break mdast structure 1`] = `
"✨新增

<details>
<summary>正式版获得 v2.3.1 ~ v2.3.3 预览版的功能</summary>

*   \`清爽首页\` 的热门视频支持显示弹幕数量.
*   \`极简首页\` 初版已完成.
*   脚本的更新源配置默认值更换为 \`GitHub\`, 并添加了 \`AltCdn\`, 表示开发者自定义的其他 CDN 源. \`jsDelivr\` 将会删除.
*   在 \`GitHub\` 更新源下, MDI 图标库更换使用 GitHub Pages.
*   新增组件 \`UP 主黑名单\` by [snowraincloud](https://github.com/snowraincloud). (PR #3537)
*   在无限滚动的场景下, 点击那个 \`加载中\` 的标识 (ScrollTrigger) 可以手动触发加载下一页. 在遇到没有自动加载下一页的情况时会比较有用.
*   \`下载视频\` 支持 \`flac\` 音源. (#3497)
*   \`自定义顶栏\` 支持设置顶栏的高度, 设置为 64px 即为原版顶栏的高度. (#3136)

</details>

*   \`删除视频弹窗\` 支持 3.x 播放器的 \`关联视频\` 和 \`评分\` 弹窗. (#3545)
*   \`展开弹幕列表\` 支持设置 \`最大弹幕数量\`, 超过此数量不进行展开, 避免展开时卡死页面. (#2972)
*   \`图片批量导出\` 在导出动态的图片时, 可以指定 \`originalUser\` 作为被转发用户名. (#1208)
*   新增插件 \`自定义顶栏 - 频道\`, 为自定义顶栏添加一个频道入口. (#3258)
*   \`自定义顶栏\` 的历史弹窗中支持暂停/继续记录历史. (#3303)

✨新增

<details>
<summary>正式版获得 v2.1.9 ~ v2.1.10 预览版的功能</summary>

*   设置面板移动了搜索框的位置, 添加了检查更新和卸载组件的快捷按钮. (PR #3279 by [FoundTheWOUT](https://github.com/FoundTheWOUT))
*   \`自定义顶栏\` 支持硬核 LV6 会员的图标显示. (#3203)
*   \`动态过滤器\` 支持屏蔽发送动态的面板. (#2447)
*   新增插件 \`下载视频 - 手动输入\`, 可以手动输入 av / BV 号来进行下载. (#3227)
*   Toast 消息能够显示关闭时间的倒计时进度, 且鼠标进入时停止倒计时. (#3204)

</details>
"
`;

exports[`Unit Test replace RemoteContent > fetch remote content, and select two section 1`] = `
"### Sugaring time begins in February and lasts for two to six weeks

well into March. Weather permitting, Grandad always started tapping on Washington's Birthday. The sugaring season, as we called it, required thawing days and freezing nights to allow the sap to flow. Rain spoiled the sap and wind dried up the tap holes so it was a fair-weather industry. Grandad hand-drilled his tap holes using a No. 8 wood bit, drilling about 1/2 inch deep into the vein of a sugar maple tree. Into this hole, a wood spile would then be tapped and after drilling another, a pail was hung to collect the sap as it steadily dripped from the spile. The wooden spiles were about eight inches in length and three-quarters of an inch in diameter. Ours were made from elderberry as that scrub has a soft center that could be cleaned out to make a tube. Our maple grove was along the brook and the edge of the field about a have mile up from the house. There were approximately 200 sugar maples large enough to tap on Grandad's farm and the scars on the trunks of many of the older trees indicated the number of years they had been supplying sap.

## Twice a day, and some days more often

the sap was collected from each tree and carried up to the saphouse. The kettles were filled and, with a good hot fire beneath, thirty pails of sap would boil down to about eight quarts of syrup in four to five hours. As the sap boiled down and thickened it was removed from the fire and strained and then placed in the brass kettle over a low fire for the last bit of boiling down to syrup. The syrup, as it came from the saphouse, was usually not quite thick enough so it was Grandma's job to boil it a little longer on the back of the old wood-burning kitchen stove. For a thicker cream sugar (we called it \\"slush\\") and for sugar cakes, additional boiling and beating were required with all hands taking a turn. The beating cooled the heavy syrup and as it cooled it became granular and almost white. Our reward was the pan to \\"lick\\" out and hot syrup poured out on the snow for taffy.

### My brothers, sisters, and I, as youngsters

were at the saphouse every chance we got during the sugaring season, playing in the brook and jumping from rock to rock. If you slipped and got wet, the fire was always handy. When we grew older we helped to gather the sap, split wood, and tend the fire. After Grandad's death, I did the whole bit myself for a few years.
"
`;

exports[`Unit Test replace RemoteContent > html image self close 1`] = `
"<div>
  <img height=\\"400\\" src=\\"https://user-images.githubusercontent.com/26504152/242905984-895cb72c-b344-40c3-91a0-2a6b20d5f783.png\\" />
  <img height=\\"400\\" src=\\"https://user-images.githubusercontent.com/26504152/242905640-cbc948f1-734e-46f2-96a7-d57787b7cf47.png\\" />
</div>
"
`;

exports[`Unit Test replace RemoteContent > html tag 1`] = `
"# 合集包

合集包提供了批量的功能安装链接, 方便一次性安装大量功能.

#### 简洁至上

简化各种多余界面元素, 专注于内容本身.

包含以下功能:
删除广告, 删除直播水印, 删除视频弹窗, 禁用特殊弹幕样式, 简化评论区, 简化直播间, 简化首页, 自动收起直播侧栏, 隐藏视频推荐, 隐藏直播推荐, 隐藏视频标题层, 自动隐藏侧栏

<details>
<summary><strong>jsDelivr Stable</strong></summary>
https://cdn.jsdelivr.net/gh/the1812/Bilibili-Evolved@v2/registry/dist/components/utils/remove-promotions.js
https://cdn.jsdelivr.net/gh/the1812/Bilibili-Evolved@v2/registry/dist/components/live/remove-watermark.js
https://cdn.jsdelivr.net/gh/the1812/Bilibili-Evolved@v2/registry/dist/components/video/player/remove-popup.js
https://cdn.jsdelivr.net/gh/the1812/Bilibili-Evolved@v2/registry/dist/components/style/special-danmaku.js
https://cdn.jsdelivr.net/gh/the1812/Bilibili-Evolved@v2/registry/dist/components/style/simplify/comments.js
https://cdn.jsdelivr.net/gh/the1812/Bilibili-Evolved@v2/registry/dist/components/style/simplify/live.js
https://cdn.jsdelivr.net/gh/the1812/Bilibili-Evolved@v2/registry/dist/components/style/simplify/home.js
https://cdn.jsdelivr.net/gh/the1812/Bilibili-Evolved@v2/registry/dist/components/live/side-bar.js
https://cdn.jsdelivr.net/gh/the1812/Bilibili-Evolved@v2/registry/dist/components/style/hide/video/related-videos.js
https://cdn.jsdelivr.net/gh/the1812/Bilibili-Evolved@v2/registry/dist/components/style/hide/video/recommended-live.js
https://cdn.jsdelivr.net/gh/the1812/Bilibili-Evolved@v2/registry/dist/components/style/hide/video/top-mask.js
https://cdn.jsdelivr.net/gh/the1812/Bilibili-Evolved@v2/registry/dist/components/style/auto-hide-sidebar.js
</details>

### Sugaring time begins in February and lasts for two to six weeks

well into March. Weather permitting, Grandad always started tapping on Washington's Birthday. The sugaring season, as we called it, required thawing days and freezing nights to allow the sap to flow. Rain spoiled the sap and wind dried up the tap holes so it was a fair-weather industry. Grandad hand-drilled his tap holes using a No. 8 wood bit, drilling about 1/2 inch deep into the vein of a sugar maple tree. Into this hole, a wood spile would then be tapped and after drilling another, a pail was hung to collect the sap as it steadily dripped from the spile. The wooden spiles were about eight inches in length and three-quarters of an inch in diameter. Ours were made from elderberry as that scrub has a soft center that could be cleaned out to make a tube. Our maple grove was along the brook and the edge of the field about a have mile up from the house. There were approximately 200 sugar maples large enough to tap on Grandad's farm and the scars on the trunks of many of the older trees indicated the number of years they had been supplying sap.
"
`;

exports[`Unit Test replace RemoteContent > should return mdx properly 1`] = `
"import Comp from \\"./Comp\\";

<Comp />

<Comp>123</Comp>

## h2

body
"
`;
